src/webui/api/torrentscontroller.cpp

```diff
@@ -1619,3 +1787,248 @@ void TorrentsController::setSSLParametersAction()
+            qDebug("Fetching metadata for %s", qUtf8Printable(infoHash.toTorrentID().toString()));
+            if (!session->downloadMetadata(torrentDescr)) [[unlikely]]
+            {
+                // our checks above should prevent ever hitting this
+                qDebug("Unable to fetch metadata for %s", qUtf8Printable(infoHash.toTorrentID().toString()));
+                throw APIError(APIErrorType::BadParams, tr("Unable to download metadata for '%1'").arg(infoHash.toTorrentID().toString()));
+            }
+
+            m_torrentSource.insert(source, infoHash);
+            m_torrentMetadata.insert(infoHash, torrentDescr);
+
+            setStatus(APIStatus::Async);
+            setResult(serializeInfoHash(infoHash));
+        }
+
+        return;
+    }
+
+    throw APIError(APIErrorType::BadParams, tr("Unable to parse '%1'").arg(source));
+}
+
+void TorrentsController::parseMetadataAction()
+{
+    const DataMap &uploadedTorrents = data();
+    // must provide some value to parse
+    if (uploadedTorrents.size() == 0)
+        throw APIError(APIErrorType::BadParams, tr("Must specify torrent file(s)"));
+
+    QJsonObject result;
+    for (auto it = uploadedTorrents.constBegin(); it != uploadedTorrents.constEnd(); ++it)
+    {
+        if (const auto loadResult = BitTorrent::TorrentDescriptor::load(it.value()))
+        {
+            const BitTorrent::TorrentDescriptor &torrentDescr = loadResult.value();
+            const BitTorrent::InfoHash infoHash = torrentDescr.infoHash();
+
+            const QString &fileName = it.key();
+            m_torrentSource.insert(FILE_SOURCE_PREFIX + fileName, infoHash);
+            m_torrentMetadata.insert(infoHash, torrentDescr);
+
+            result.insert(fileName, serializeTorrentInfo(torrentDescr));
+            continue;
+        }
+
+        throw APIError(APIErrorType::BadData, tr("'%1' is not a valid torrent file.").arg(it.key()));
+    }
+
+    setResult(result);
+}
+
+void TorrentsController::saveMetadataAction()
+{
+    requireParams({u"source"_s});
+
+    const QString sourceParam = params()[u"source"_s].trimmed();
+    if (sourceParam.isEmpty())
+        throw APIError(APIErrorType::BadParams, tr("Must specify URI or hash"));
+
+    const QString source = QUrl::fromPercentEncoding(sourceParam.toLatin1());
+    if (const auto iter = m_torrentSource.find(source); iter != m_torrentSource.end())
+    {
+        const BitTorrent::InfoHash &infoHash = iter.value();
+        if (isMetadataDownloaded(infoHash))
+        {
+            const BitTorrent::TorrentDescriptor &torrentDescr = m_torrentMetadata[infoHash];
+            const nonstd::expected<QByteArray, QString> result = torrentDescr.saveToBuffer();
+            if (!result)
+                throw APIError(APIErrorType::Conflict, tr("Unable to export torrent metadata. Error: %1").arg(result.error()));
+
+            setResult(result.value(), u"application/x-bittorrent"_s, (infoHash.toTorrentID().toString() + u".torrent"));
```

Comment:
We should have a global wide constant for this MIME type. Perhaps in another PR.